cmd/vm-informant/main.go

```diff
@@ -64,3 +96,55 @@ func main() {
 		klog.Fatalf("Server failed: %s", err)
 	}
 }
+
+// runRestartOnFailure repeatedly calls this binary with the same flags, but with 'auto-restart'
+// removed.
+//
+// We execute ourselves as a subprocess so that it's possible to appropriately cleanup after
+// termination by various signals (or an unhandled panic!). This is worthwhile because we *really*
+// don't want to leave the cgroup frozen while waiting to restart.
+func runRestartOnFailure(args []string, cleanupHooks []func()) {
+	selfPath := os.Args[0]
+
+	minWaitDuration := time.Second * 5
+
+	for {
+		minWait := time.After(minWaitDuration)
+
+		cmd := exec.CommandContext(context.TODO(), selfPath, args...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		err := cmd.Run()
+
+		var exitMode string
+
+		if err != nil {
+			// lint note: the linter's worried about wrapped errors being incorrect with switch, but
+			// this is cleaner than the alternative (via errors.As) and it's still correct because
+			// exec.Command.Run() explicitly mentions ExitError.
+			switch err.(type) { //nolint:errorlint // see above.
+			case *exec.ExitError:
+				exitMode = "failed"
+				klog.Errorf("vm-informant exited with: %v", err)
+			default:
+				exitMode = "failed to start"
+				klog.Errorf("error running vm-informant: %v", err)
+			}
+		} else {
+			exitMode = ""
+			klog.Warningf("vm-informant exited without error. This should not happen.")
+		}
+
+		for _, h := range cleanupHooks {
+			h()
+		}
+
+		select {
+		case <-minWait:
+			klog.Infof("vm-informant restarting immediately")
+		default:
+			klog.Infof("vm-informant %s. respecting minimum wait of %s", exitMode, minWaitDuration)
+			<-minWait
+		}
```

Comment:
This is a really complicated way of (modulo logging): collecting the start time at the top of the loop and then: 

```go
time.Sleep(time.Since(startAt)-minWaitDuration)
```

The channel business only gets you something if you're processing multiple channels in a select or if you want to allow it to cancel.

Reply:
Fixed